virtctl: Overhaul virtctl create vm docs (#851)
kubevirt/user-guide: docs/user_workloads/creating_vms.md

The virtctl sub command `create vm` allows easy creation of VirtualMachine
manifests from the command line. It leverages
[instance types and preferences](../user_workloads/instancetypes.md) and
inference by default (see
[Using instance types and preferences](#using-instance-types-and-preferences))

> and provides several flags to control details of the created virtual machine.

Reviewer: ```suggestion
and it provides several flags to control details of the created virtual machine.
```

virtualmachine.kubevirt.io/my-vm created
```

## Using instance types and preferences

> Instance types and preferences can be used with the appropriate flags. If
> not otherwise specified, instance types and preferences are inferred from

Reviewer: ```suggestion
Instance types and preferences can be used with the appropriate flags. If they are not specified, instance types and preferences are inferred from
```

If a prefix was not supplied the cluster scoped resources will be used by
default.

### Inference of instance type and/or preference

To explicitly

> infer [instance types and/or preferences](./instancetypes.md#inferFromVolume)

Author: Didn't think of that. I reverted this change.

from the volume used to boot the virtual machine add the following flags:

```shell
virtctl create vm --infer-instancetype --infer-preference
```

> The implicit default is to always try inferring an instance type and

Reviewer: ```suggestion
The implicit default is to always try to infer an instance type and
```
I'm not 100% sure about this...

preference from the boot volume. This feature makes use of the
`IgnoreInferFromVolumeFailure` policy, which suppresses failures on inference

> of instance types and preferences. If one of the above switches was provided

Reviewer: ```suggestion
of instance types and preferences. If one of the above switches have been explicitly provided
```

explicitly specified, the `RejectInferFromVolumeFailure` policy is used
instead. This way users are made aware of potential issues during the

> virtual machinecreation.

Reviewer: ```suggestion
virtual machine creation.
```

boot order regardless of the order their flags were specified on the
command line:

1. Containerdisks
2. Directly used PVCs
3. DataSources
4. Cloned PVCs
5. Blank volumes
6. Imported volumes (through the `--volume-import` flag)

If multiple volumes of the same kind were specified their order is
determined by the order in which their flags were specified.

## Generating cloud-init user data

> To generate cloud-init user data with `virtctl create vm` the following
> flags can be used. Generating cloud-init user data is mutually exclusive
> with [specifying custom cloud-init user data](#specifying-custom-cloud-init-user-data),
> as explained below.

Reviewer: ```suggestion
To generate cloud-init user data with `virtctl create vm` the following
flags can be used: 

> *Note:* Generating cloud-init user data is mutually exclusive
> with [specifying custom cloud-init user data](#specifying-custom-cloud-init-user-data),
> as explained below.
```

### `--user` flag

This flag allows to specify the main user of the virtual machine that is
created by cloud-init. It sets the `user` parameter in the generated
cloud-init user data.

### `--password-file` flag

This flag allows to specify a file to read the password for the virtual

> machine's main user from. In the generated cloud-init user data it sets the

Reviewer: ```suggestion
machine's main user from. In the generated cloud-init user data, it sets the
```

value of the `password` parameter to the read in value and the value of the
`chpasswd` parameter to `{ expire: False }`.

### `--ssh-key` flag

This flag allows to specify one or more SSH authorized keys for the virtual
machine's main user. It sets the `ssh_authorized_keys` parameter in the
generated cloud-init user data.

### `--ga-manage-ssh` flag

When this flag is set a command enabling the `qemu-guest-agent` to manage SSH
authorized keys is added to the generated cloud-init user data. The command

> is added to the `runcmd` parameter and is required on SELinux enabled

Reviewer: ```suggestion
is added to the `runcmd` parameter which is required on SELinux enabled
```

This flag allows to specify a file to read the password for the virtual
machine's main user from. In the generated cloud-init user data, it sets the
value of the `password` parameter to the read in value and the value of the
`chpasswd` parameter to `{ expire: False }`.

### `--ssh-key` flag

This flag allows to specify one or more SSH authorized keys for the virtual
machine's main user. It sets the `ssh_authorized_keys` parameter in the
generated cloud-init user data.

### `--ga-manage-ssh` flag

When this flag is set a command enabling the `qemu-guest-agent` to manage SSH
authorized keys is added to the generated cloud-init user data. The command
is added to the `runcmd` parameter which is required on SELinux enabled
distributions that would otherwise not allow the `qemu-guest-agent` to manage
SSH authorized keys in the home directories of users.

### Example

```shell
$ virtctl create vm --user myuser --ssh-key THISISMYKEY --ga-manage-ssh
```

This command will generate the following cloud-init user data:

```yaml
volumes:
  - cloudInitNoCloud:
      userData: |-
        #cloud-config
        user: myuser
        ssh_authorized_keys:
          - THISISMYKEY
        runcmd:

>           - [ setsebool, -P, 'virt_qemu_ga_manage_ssh', 'on' ]

Reviewer 2: This is slightly confusing as `--ga-manage-ssh` isn't useful when a user provides `--ssh-key` right?

I'd break this out into another example or just explicitly use it in the access credentials example below.

Author: Ah yeah, well... not sure what I was thinking. This should be using `--access-cred`.

boot order regardless of the order their flags were specified on the
command line:

1. Containerdisks
2. Directly used PVCs
3. DataSources
4. Cloned PVCs
5. Blank volumes
6. Imported volumes (through the `--volume-import` flag)

If multiple volumes of the same kind were specified their order is
determined by the order in which their flags were specified.

## Generating cloud-init user data

To generate cloud-init user data with `virtctl create vm` the following
flags can be used.

!!! Note
    Generating cloud-init user data is mutually exclusive with [specifying custom cloud-init user data](#specifying-custom-cloud-init-user-data), as explained below.

### `--user` flag

> This flag allows to specify the main user of the virtual machine that is

Author: I'm not a man of too many words! ACK

virtctl create vm --cloud-init-user-data $CLOUD_INIT_USERDATA
```

## Adding access credentials to a virtual machine

> By using the `--access-cred` flag `virtctl create vm` can configure

Reviewer: ```suggestion
By using the `--access-cred` flag, the `virtctl create vm` command can configure
```

[access credentials](./accessing_virtual_machines.md#dynamic-ssh-public-key-injection-via-qemu-guest-agent)
in a created virtual machine. It supports SSH authorized key and password
access credentials and can configure them to be injected either through the
`qemu-guest-agent` or through cloud-init metadata. The supported parameters of
the flag depend on the chosen `type` and `method`. The flag can be passed
multiple times to configure more than one access credential.

This flag interacts with the flags used to
[generate cloud-init user data](#generating-cloud-init-user-data), namely it
inherits the same `--user` for SSH key injection, and it enables

> `qemu-guest-agent` to manage SSH authorized keys (`--ga-manage-ssh`), if not
> explicitly disabled by the user.

Reviewer: ```suggestion
`qemu-guest-agent` to manage SSH authorized keys (`--ga-manage-ssh`), if it is not
explicitly disabled by the user.
```

### Example

```shell
$ virtctl create vm --user myuser --access-cred=src:my-keys
```

This command will generate the following access credentials and cloud-init user
data:

```
[...]
accessCredentials:
  - sshPublicKey:
      propagationMethod:
        qemuGuestAgent:
          users:
            - myuser
      source:
        secret:
          secretName: my-keys
volumes:
  - cloudInitNoCloud:
      userData: |-
        #cloud-config
        user: myuser
        runcmd:
          - [ setsebool, -P, 'virt_qemu_ga_manage_ssh', 'on' ]
    name: cloudinitdisk
```

## Adding a sysprep volume

A [sysprep volume](startup_scripts.md#sysprep-examples) can be added to

> created virtual machines by passing the `--volume-sysprep` flag to `virtctl
> create vm`.

Reviewer: ```suggestion
created virtual machines by passing the `--volume-sysprep` flag to the `virtctl create vm` command.
```

```shell
virtctl create vm --memory=1Gi \
  --volume-containerdisk=src:my.registry/my-image:my-tag
```

Create a manifest for a VirtualMachine with a cloned DataSource in namespace and
specified size:

```shell
virtctl create vm --volume-import=type:ds,src:my-ns/my-ds,size:50Gi
```

Create a manifest for a VirtualMachine with a cloned DataSource and inferred
instance type and preference:

```shell
virtctl create vm --volume-import=type:ds,src:my-annotated-ds \
  --infer-instancetype --infer-preference
```

Create a manifest for a VirtualMachine with multiple volumes and specified boot
order:

```shell
virtctl create vm --volume-containerdisk=src:my.registry/my-image:my-tag \
  --volume-import=type:ds,src:my-ds,bootorder:1
```

Create a manifest for a VirtualMachine with multiple volumes and inferred
instance type and preference with specified volumes:

```shell
virtctl create vm --volume-import=type:ds,src:my-annotated-ds \
  --volume-pvc=my-annotated-pvc --infer-instancetype=my-annotated-ds \
  --infer-preference=my-annotated-pvc
```

Create a manifest for a VirtualMachine with a specified
VirtualMachineCluster{Instancetype,Preference} and cloned PVC:

```shell

> virtctl create vm --volume-import=type:pvc,src:my-ns/my-pvc

Author: Thanks for spotting this, that was obviously wrong.

virtctl create vm --preference=virtualmachinepreference/my-preference
```

Create a manifest for a VirtualMachine with specified memory and an ephemeral
containerdisk volume

```shell
virtctl create vm --memory=1Gi \
  --volume-containerdisk=src:my.registry/my-image:my-tag
```

Create a manifest for a VirtualMachine with a cloned DataSource in namespace and
specified size

```shell
virtctl create vm --volume-import=type:ds,src:my-ns/my-ds,size:50Gi
```

Create a manifest for a VirtualMachine with a cloned DataSource and inferred

> instance type and preference

Reviewer: ```suggestion
instance type and preference:
```

```shell
virtctl create vm --volume-import=type:ds,src:my-annotated-ds \
  --infer-instancetype --infer-preference
```

Create a manifest for a VirtualMachine with multiple volumes and specified boot

> order

Reviewer: ```suggestion
order:
```

```shell
virtctl create vm --volume-containerdisk=src:my.registry/my-image:my-tag \
  --volume-import=type:ds,src:my-ds,bootorder:1
```

Create a manifest for a VirtualMachine with multiple volumes and inferred

> instance type and preference with specified volumes

Reviewer: ```suggestion
instance type and preference with specified volumes:
```

```shell
virtctl create vm --volume-import=type:ds,src:my-annotated-ds \
  --volume-pvc=my-annotated-pvc --infer-instancetype=my-annotated-ds \
  --infer-preference=my-annotated-pvc
```

Create a manifest for a VirtualMachine with a specified

> VirtualMachineCluster{Instancetype,Preference} and cloned PVC

Reviewer: ```suggestion
VirtualMachineCluster{Instancetype,Preference} and cloned PVC:
```

```shell
virtctl create vm --volume-import=type:pvc,src:my-ns/my-pvc
```

Create a manifest for a VirtualMachine with a specified

> VirtualMachineCluster{Instancetype,Preference} and directly used PVC

Reviewer: ```suggestion
VirtualMachineCluster{Instancetype,Preference} and directly used PVC:
```

```shell
virtctl create vm --volume-pvc=src:my-pvc
```

Create a manifest for a VirtualMachine with a clone DataSource and a blank

> volume

Reviewer: ```suggestion
volume:
```

```shell
virtctl create vm --volume-import=type:ds,src:my-ns/my-ds \
  --volume-import=type:blank,size:50Gi
```

Create a manifest for a VirtualMachine with a specified

> VirtualMachineCluster{Instancetype,Preference} and cloned DataSource

Reviewer: ```suggestion
VirtualMachineCluster{Instancetype,Preference} and cloned DataSource:
```

```shell
virtctl create vm --instancetype=my-instancetype --preference=my-preference \
  --volume-import=type:ds,src:my-ds
```

Create a manifest for a VirtualMachine with a specified
VirtualMachineCluster{Instancetype,Preference} and two cloned DataSources (flag

> can be provided multiple times)

Reviewer: ```suggestion
can be provided multiple times):
```

```shell
virtctl create vm --instancetype=my-instancetype --preference=my-preference \
  --volume-import=type:ds,src:my-ds1 --volume-import=type:ds,src:my-ds2
```

Create a manifest for a VirtualMachine with a specified

> VirtualMachineCluster{Instancetype,Preference} and directly used PVC

Reviewer: ```suggestion
VirtualMachineCluster{Instancetype,Preference} and directly used PVC:
```

```shell
virtctl create vm --instancetype=my-instancetype --preference=my-preference \
  --volume-pvc=my-pvc
```

> Create a manifest for a VirtualMachine with a specified DataVolumeTemplate

Reviewer: ```suggestion
Create a manifest for a VirtualMachine with a specified DataVolumeTemplate:
```

```shell
virtctl create vm \
  --volume-import=type:pvc,name:my-pvc,namespace:default,size:256Mi
```

Create a manifest for a VirtualMachine with a generated cloud-init config

> setting the user and adding an ssh authorized key

Reviewer: ```suggestion
setting the user and adding an ssh authorized key:
```

```shell
virtctl create vm --user=cloud-user --ssh-key="ssh-ed25519 AAAA...."
```

Create a manifest for a VirtualMachine with a generated cloud-init config

> setting the user and setting the password from a file

Reviewer: ```suggestion
setting the user and setting the password from a file:
```

```shell
virtctl create vm --user=cloud-user --password-file=/path/to/file
```

Create a manifest for a VirtualMachine with SSH public keys injected into the VM

> from a secret called my-keys to the user also specified in the cloud-init config

Reviewer: ```suggestion
from a secret called my-keys to the user also specified in the cloud-init config:
```

```shell
virtctl create vm --user=cloud-user --access-cred=type:ssh,src:my-keys
```

Create a manifest for a VirtualMachine with SSH public keys injected into the VM

> from a secret called my-keys to a user specified as param

Reviewer: ```suggestion
from a secret called my-keys to a user specified as param:
```

```shell
virtctl create vm --access-cred=type:ssh,src:my-keys,user:myuser
```

Create a manifest for a VirtualMachine with password injected into the VM from a

> secret called my-pws

Reviewer: ```suggestion
secret called my-pws:
```

```shell
virtctl create vm --access-cred=type:password,src:my-pws
```

Create a manifest for a VirtualMachine with a Containerdisk and a Sysprep

> volume (source ConfigMap needs to exist)

Reviewer: ```suggestion
volume (source ConfigMap needs to exist):
```
